includes/Modules/Tag_Manager/Settings.php

```diff
@@ -58,4 +59,47 @@ protected function get_default() {
 			'useSnippet'     => true,
 		);
 	}
+
+	/**
+	 * Gets the callback for validating the setting's value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable|null
+	 */
+	protected function get_validate_callback() {
+		return function( $option ) {
+			$required_fields = array( 'accountID' );
+
+			// 'usageContext' is not an actual sub-setting, but can be specified to modify requirements.
+			if ( isset( $option['usageContext'] ) && Tag_Manager::USAGE_CONTEXT_WEB === $option['usageContext'] ) {
+				$required_fields[] = 'containerID';
+			} elseif ( isset( $option['usageContext'] ) && Tag_Manager::USAGE_CONTEXT_AMP === $option['usageContext'] ) {
+				$required_fields[] = 'ampContainerID';
+			} else { // Otherwise, both container fields are required, e.g. for secondary AMP.
+				$required_fields[] = 'containerID';
+				$required_fields[] = 'ampContainerID';
+			}
```

Comment:
I'm not sure about the idea of having variable required fields in this way (using non-option keys as context for validation). I don't think Settings should be `usageContext` aware. This logic belongs closer to the controller I think, as it depends on the request, or amp mode from `Context`.

In this case, I think this should happen in the REST route's `validate_callback`, where it has access to both, as well as the module instance itself. It also makes more sense to return a `WP_Error` in the context of a REST request (route/datapoint) handler than a settings class (a validation error could be an exception as well) I think.

I like the idea of `settings->validate()` but I think this shows that the Settings instance lacks enough context to do more than primitive validation. What do you think about calling `validate_required_fields` instead? This way the `$required_fields` could be provided by the context-aware consumer, such as datapoint handler. Not sure exactly how it would work for the new REST route, but I'd be interested to hear your thoughts.

Reply:
Agreed, I wasn't too happy with this so far. I guess we will have a little bit of duplication between REST-specific validation and the more generic setting validation.

Specifically for this one though: I think we should actually get rid of the `usageContext` parameter. It's pointless, because that parameter must always reference the current "AMP mode" anyway. Requiring different fields here based on the AMP mode is reasonable I think, and that would be okay to be handled by the `Module_Settings` instance itself.